scarf/readers.py

```diff
@@ -382,6 +392,86 @@ def _read_dataset(self, key: Optional[str] = None):
             vals = None
         return vals
 
+    def read_header(self) -> pl.DataFrame:
+        header = pl.read_csv(
+            self.matFn,
+            skip_rows=2,
+            separator=self.sep,
+            has_header=False,
+            n_rows=1,
+            new_columns=["nFeatures", "nCells", "nCounts"],
+        )
+        return header
+
+    def process_batch(self, dfs: pl.DataFrame, filtering_cutoff: int) -> List:
+        """Returns a list of valid barcodes after filtering out background barcodes for a given batch.
+
+        Args:
+            dfs: A Polar DataFrame containing a chunk of data from the MTX file.
+            filtering_cutoff: The cutoff value for filtering out background barcodes
+        """
+        dfs_ = dfs.group_by('barcode').agg(pl.sum('count'))
+        dfs_ = dfs_.filter(pl.col('count') > filtering_cutoff)
+        return np.sort(dfs_['barcode'])
+
+    def _get_valid_barcodes(
+        self, filtering_cutoff: int,
+        batch_size: int = int(10e4),
+        lines_in_mem: int = int(10e6)
+    ) -> np.ndarray:
+        """Returns a list of valid barcodes after filtering out background barcodes.
+
+        Args:
+            filtering_cutoff: The cutoff value for filtering out background barcodes.
+            batch_size: The number of barcodes to process in each batch.
+            lines_in_mem: The number of lines to read into memory
+        """
+        test_counter = 0
+        matrixIO = pl.scan_csv(
+            self.matFn, skip_rows=3, separator=self.sep, has_header=False,
+        )
+        assert len(matrixIO.collect_schema().names()) == 3
+        matrixIO = matrixIO.rename({'column_1': 'gene', 'column_2': 'barcode', 'column_3': 'count'})
+        header = self.read_header()
+        nChunks = math.ceil(header["nCounts"][0] / lines_in_mem)
+        test_counter = 0
+        valid_idx = []
+        start = 1
+        dfs = pl.DataFrame()
+        for i in tqdmbar(
+            range(nChunks), desc="Filtering out background barcodes"
+        ):
+            chunk = matrixIO.slice(i*lines_in_mem, lines_in_mem).collect()
+            # Check if we've reached or exceeded the current batch boundary
+            if (chunk[-1]['barcode'][0] - start) >= batch_size: # If the last "cell id" is greater than the start + batch size
+                # Filter rows in the current chunk that belong to the current batch
+                idx = np.array(chunk['barcode'] < (batch_size + start))  # This is the crucial line. This makes sure that if any cell ID is spread over multiple chunks, it is not missed, as any cell ID that is less than the batch size + start is included.
```

Comment:
I'm not well versed in Python dataframe syntax;  are we depending on the barcodes being ordered? Not sure that holds true.

Reply:
Yes, the barcodes are ordered in ascending order in the `matrix.mtx` file. The order may not be continuous in raw, (for example, it may start from 1 and next entry as 23) but ascending for sure.

Reply:
Ah right, it’s all indexes in the mtx file